### ISSUE TITLE
Handle very large tasks in a scalable manner

### ISSUE THREAD

**Issue by espenak** [devilry/devilry-django#572]
We have some tasks that we can optimize to a point where running them in realtime should not be a problem. But some tasks require a lot of time, and some tasks scale badly with a lot of concurrent users.

The biggest problem is write heavy tasks. ACID compiant databases only scale such tasks to a certain point because they have to write and flush the data to disk before they can be considered  completed. It is possible to mitigate this problem by throwing more hardware at the problem (using a database cluster), but that should not be required just to handle the extreme cases.

Most read heavy tasks can be optimized by smart caching and task specific database optimizations (storing preprocessed data in cache fields in the database). But when we work with huge enough datasets, even such optimizations fall short.
## The different types of views
- **Bulk create views**: Requires a load spinner with a message that tells any user visiting the page that some functionality is unavailable while some objects are beeing created (the text must be view specific).
- **Report views**: Views that has to fetch a lot of data before the view can be shown.
## Frontend solution

A load spinner on all pages that is not available while the data is beeing created. The load spinner must be connected to an API that we can poll at some interval to check if the task is complete. We should start polling every second, but gradually change this to every 5 second.
## Backend solution

We should have a database model, BatchTask, where we register bulk tasks with:
- `content_object`: a GenericForeignKey to some object that the tasks is beeing performed on. The object is typically the parent object, like the Assignment if the task is adding AssignmentGroups to an Assignment.
- `started_by`: A ForeignKey to a User object.
- `tasktype`: A unique string for this type of task. Makes it possible to distinguish different background tasks on the same object. For Assignment, one tasktype may be `addgroups` and another may be `createstatisticspdf`.
- `start_time`: The time when the task started.
- `end_time`: The time when the task ended. `None` if still running.
- `successful`: `True` if the task finished successfully. `None` if still running.
- `errorcode`: A string/tag describing the error (if successful is False). This should be something that is more readable that a numeric error code, and something that we can easily use to determine what to tell the user.

When a task starts, we create a `BatchTask` object, and invoke a Celery task with the BatchTask as one of the inputs. When the Celery task completes, it updates the BatchTask.

We should have a REST API that can be polled to get the status of a task. The parameters are the natural key (the id and type of the content_object) and the tasktype. The output is all the information about the BatchTask except `started_by`. This API should only require login. Security is not a problem since the API gives no private information.
## Frontend solution - more details

The frontend solution should probably be as simple as the following Django template code:

``` django
{% batchtask_loadspinner with content_object=someobject %}
{% batchtask_loadspinner_loading %}
    <i class="fa fa-spinner fa-spin"></i>
    Creating something...
{% batchtask_loadspinner_success %}
        Success!
{% batchtask_loadspinner_error %}
    Something went wrong!
    {% if errorcode=='someerror' %}
        Some error:(
    {% else %}
        Unknown error!!
    {% endif %}
{% endbatchtask_loadspinner %}
```

The `batchtask_loadspinner` tag should take care of the javascript to poll for requests, and it should take care of showing/hiding the various sections. Not sure that this is the most natural template tag structure, but I think it should look something like this.
## Where to add this extra complexity?

On pages where users experience that the UI times out. We should also analyse our pages and consider adding this in places where timeouts is likely, such as:
- The overview of all results in a semester.
- The qualified for final exams wizard.
- Add candidates to assignment? May not be needed after #554.
- Build complex statistics

**Comment by torgeirl** [comment]
No longer relevant.
